### DISCUSSION TITLE
Jetpack 5 and custom partition  layout

### DISCUSSION THREAD

**Discussion by hannes09** [OE4T/meta-tegra#1136] [Q&A] [answered]
I want to use a custom partiton layout and have added the following entry to my _custom_flash_layout.xml_.

```
        <partition name="custom_data_part" type="data">
            <allocation_policy> sequential </allocation_policy>
            <filesystem_type> basic </filesystem_type>
            <size>803316480 </size>
            <file_system_attribute> 0 </file_system_attribute>
            <allocation_attribute> 0x8 </allocation_attribute>
            <align_boundary> 4096 </align_boundary>
            <percent_reserved> 0 </percent_reserved>
            <filename> app.ext4.img </filename>
        </partition>
```

The _custom_flash_layout.xml_ is almost the same as _flash_t194_sdmmc.xml_
The custom_data_part was added in front of the data partition.
This approch has worked with JP 4.6

With JP 5.0.2 (kirkstone branch) I run into an error:

```
| [   1.3112 ] partition_id	partition_name			      StartingLba	EndingLba
| [   1.3113 ]            1	BCT                                 	       0	      63
| [   1.3114 ]            2	mb1                                 	      64	     575
| [   1.3115 ]            3	mb1_b                               	     576	    1087
| [   1.3115 ]            4	MB1_BCT                             	    1088	    1215
| [   1.3115 ]            5	MEM_BCT                             	    1216	    1663
| [   1.3115 ]            6	spe-fw                              	    1664	    2175
| [   1.3115 ]            7	mb2                                 	    2176	    2623
| [   1.3115 ]            8	mts-preboot                         	    2624	    2687
| [   1.3115 ]            9	reserved_for_chain_A_boot           	    2688	    6783
| [   1.3115 ]           10	MB1_BCT_b                           	    6784	    6911
| [   1.3115 ]           11	MEM_BCT_b                           	    6912	    7359
| [   1.3115 ]           12	spe-fw_b                            	    7360	    7871
| [   1.3115 ]           13	mb2_b                               	    7872	    8319
| [   1.3115 ]           14	mts-preboot_b                       	    8320	    8383
| [   1.3115 ]           15	reserved_for_chain_B_boot           	    8384	   12479
| [   1.3115 ]           16	BCT-boot-chain_backup               	   15616	   15743
| [   1.3115 ]           17	reserved_partition                  	   15744	   15871
| [   1.3115 ]           18	secondary_gpt_backup                	   15872	   15999
| [   1.3115 ]           19	VER_b                               	   16000	   16127
| [   1.3115 ]           20	VER                                 	   16128	   16255
| [   1.3115 ] End sector for custom_data_part, expected at: 61579518, actual: 0
| [   1.3115 ]
```

Can you give an example how to add a custom partition to the xml file?

**Comment by madisongh** [comment, marked as answer]
That error indicates that you have exceeded the available space in the flash.  There is less free space available with JetPack 5 vs. JetPack 4, so you may have to reduce the `ROOTFSPART_SIZE` setting (which controls how large the APP partition is) to accommodate the space you want to use for the added partition.

**Reply by hannes09** [reply]
I have adjusted ROOTFSPART_SIZE but it does not solve the problem.

**Reply by madisongh** [reply]
Hmm, well, it looks like NVIDIA, for some reason, set fixed starting locations for some partitions in `flash_t194_sdmmc.xml` in R35.1.0.

One is the `mts-mce_b` partition, that has
```
            <start_location> 0x708400000 </start_location> <!-- aligned to 0x100000 -->
```
Another is the recovery kernel partition ("RECNAME" in the original file, which gets rewritten as "recovery" during processing), which has:
```
            <start_location> 0x710800000 </start_location> <!-- aligned to 0x100000 -->
```
So reducing the size of APP won't make room for partitions you add after either of those two.  I don't know why NVIDIA chose to do this; the comments there imply it's just for aligning those partitions to a 1MiB boundary.  A forum post might be called for to ask them.

But to resolve this, you could either relocate your added partition to just after APP, instead of at the end.  Alternatively, you could try changing those `<start_location>` elements to be:
```
	    <align_boundary> 0x100000 </align_boundary>
```
which will force the alignment mentioned in the comments without specifying a fixed starting point.  This second option is potentially riskier, since I don't know whether the fixed locations in the original file are there for a good reason or not.

Either way, you'll still need to ensure you've reduced `ROOTFSPART_SIZE` to allow for the added partition to fit.
